contracts/exchange/contracts/src/MixinAssetProxyDispatcher.sol

```diff
@@ -142,31 +158,69 @@ contract MixinAssetProxyDispatcher is
                 mstore(add(cdStart, 68), and(to, 0xffffffffffffffffffffffffffffffffffffffff))
                 mstore(add(cdStart, 100), amount)
-                
+
                 /////// Setup Data Area ///////
                 // This area holds `assetData`.
                 let dataArea := add(cdStart, 132)
+                let assetDataArea := assetData
                 // solhint-disable-next-line no-empty-blocks
                 for {} lt(dataArea, cdEnd) {} {
-                    mstore(dataArea, mload(assetData))
+                    mstore(dataArea, mload(assetDataArea))
                     dataArea := add(dataArea, 32)
-                    assetData := add(assetData, 32)
+                    assetDataArea := add(assetDataArea, 32)
                 }
 
                 /////// Call `assetProxy.transferFrom` using the constructed calldata ///////
-                let success := call(
+                didSucceed := call(
                     gas,                    // forward all gas
                     assetProxy,             // call address of asset proxy
                     0,                      // don't send any ETH
                     cdStart,                // pointer to start of input
-                    sub(cdEnd, cdStart),    // length of input  
+                    sub(cdEnd, cdStart),    // length of input
                     cdStart,                // write output over input
-                    512                     // reserve 512 bytes for output
+                    288                     // reserve 288 bytes for output
                 )
-                if iszero(success) {
-                    revert(cdStart, returndatasize())
+
+                if iszero(didSucceed) { // Call reverted.
+                    // Attempt to to decode standard revert messages with the
+                    // signature `Error(string)` so we can upgrade them to
+                    // a rich revert.
+
+                    // Standard revert errors have the following memory layout:
+                    // | Area            | Offset  | Length   | Contents                              |
+                    // --------------------------------------------------------------------------------
+                    // | Selector        | 0       | 4        | bytes4 function selector (0x08c379a0) |
+                    // | Params          |         |          |                                       |
+                    // |                 | 4       | 32       | uint256 offset to data (o)            |
+                    // | Data            |         |          |                                       |
+                    // |                 | 4 + o   | 32       | uint256 length of data (n             |
+                    // |                 | 32 + o  | n        | Left-aligned message bytes            |
+
+                    // Make sure the return value is long enough.
+                    if gt(returndatasize(), 67) {
```

Comment:
We can. I just figured it would be a nicer, more language agnostic experience if there were less places to decode nested function signatures, because the decoding tooling wouldn't have to be as robust. `TransactionExecutionError` already does this very thing, but that's because the caught revert can be literally anything. Here we know it's always a string. 

Is the objection because of gas?

Reply:
Actually, since the asset proxies always revert a string, a lot of the checks in the assembly are unecessary. So we're just left with:
```solidity
cdStart := add(cdStart, 4) 
revertMessage := add(cdStart, mload(cdStart))
```

Reply:
Here's what [the new version](../blob/9c104f4e88b8d7dc69fc1d461a414a79865c2de2/contracts/exchange/contracts/src/MixinAssetProxyDispatcher.sol#L198) looks like:

```solidity
// Get the size of the string data/payload.
let revertDataSize := sub(returndatasize(), 36)
// We need to move the free memory pointer because we
// still have solidity logic that executes after this assembly.
mstore(64, add(cdStart, revertDataSize))
// Copy the revert string data.
returndatacopy(cdStart, 36, revertDataSize)
revertMessage := cdStart
```

Reply:
Omitting the first 36 bytes works if we assume that the revert data is always a string, but what if we move to rich revert reasons for asset proxies too? In that case, it's probably better to just include all of the return data and decode client side based off of the function selector.